packages/cdktf/lib/synthesize/synthesizer.ts

```diff
@@ -0,0 +1,38 @@
+import * as path from "path";
+import * as fs from "fs";
+import { TerraformStack } from "../terraform-stack";
+import { IStackSynthesizer, ISynthesisSession } from "./types";
+
+export class StackSynthesizer implements IStackSynthesizer {
+  constructor(protected stack: TerraformStack) {}
+
+  synthesize(session: ISynthesisSession) {
+    if (!session.skipValidation) {
+      const errors = this.stack.node.validate();
+      if (errors.length > 0) {
+        const errorList = errors
+          .map((e) => `[${this.stack.node.path}] ${e}`)
+          .join("\n  ");
+        throw new Error(
+          `Validation failed with the following errors:\n  ${errorList}`
+        );
+      }
+    }
+
+    const manifest = session.manifest;
+    const stackManifest = manifest.forStack(this.stack);
+
+    const workingDirectory = path.join(
+      session.outdir,
+      stackManifest.workingDirectory
+    );
+    if (!fs.existsSync(workingDirectory)) fs.mkdirSync(workingDirectory);
+
+    const tfConfig = this.stack.toTerraform();
```

Comment:
It just occurred to me: Since we lost `onSynthesize` and `toTerraform` is called on [Terraform Elements](https://github.com/hashicorp/terraform-cdk/blob/d904505481c7daee448b531572b2194f79b4e06f/packages/cdktf/lib/terraform-stack.ts#L157) only, we essentially don't have a way to hook into the synth event for [Resources](https://github.com/hashicorp/terraform-cdk/blob/d904505481c7daee448b531572b2194f79b4e06f/packages/cdktf/lib/resource.ts).

Reply:
Do we need to?

Reply:
It might help for https://github.com/hashicorp/terraform-cdk/blob/d904505481c7daee448b531572b2194f79b4e06f/packages/cdktf/lib/terraform-asset.ts#L26 and I've done it at least once in another project (while I'd have been able to do it differently, it was the most convenient way to achieve what I was trying to do)

Reply:
Hm, the AWS CDK v2 does not allow users to use such an API anymore it seems. There is [something](https://github.com/aws/aws-cdk/blob/dcae3eead0dbf9acb1ed80ba95bb104c64cb1bd7/packages/@aws-cdk/core/lib/app.ts#L161) to be used internally though. It looks a bit like a workaround for enabling a certain use case.

Reply:
Here is some background on that: https://github.com/aws/aws-cdk-rfcs/blob/master/text/0192-remove-constructs-compat.md#07-no-synthesize
It seems that we could re-add it ourselves, but I'd like to make sure whether we really need to.

Reply:
Re-added that but for now only exposed it internally (and used it in the Asset). If we want to open this to external usage, we should think about use cases first and whether a custom synth function would be the right way to solve them (instead of e.g. Aspects or alike).